File: src/dependent/payments/CheckoutService.java
package org.example.dependent.payments;

import java.math.BigDecimal;

public class CheckoutService {
    private final TaxCalculator taxCalculator;
    private final PaymentGatewayClient paymentGatewayClient;
    private final ReceiptGenerator receiptGenerator;

    public CheckoutService(TaxCalculator taxCalculator,
                           PaymentGatewayClient paymentGatewayClient,
                           ReceiptGenerator receiptGenerator) {
        this.taxCalculator = taxCalculator;
        this.paymentGatewayClient = paymentGatewayClient;
        this.receiptGenerator = receiptGenerator;
    }

    public Receipt processCheckout(Cart cart, String currency) {
        BigDecimal totalWithTax = taxCalculator.calculateTotalWithTax(cart, currency);
        String paymentId = paymentGatewayClient.charge(totalWithTax, currency, cart.getReference());
        return receiptGenerator.generate(paymentId, cart, totalWithTax);
    }
}

File: src/dependent/payments/checkout_service.py
from decimal import Decimal

from .tax_calculator import TaxCalculator
from .payment_gateway_client import PaymentGatewayClient
from .receipt_generator import ReceiptGenerator
from .cart import Cart
from .receipt import Receipt


class CheckoutService:
    def __init__(self, tax_calculator: TaxCalculator, payment_gateway_client: PaymentGatewayClient, receipt_generator: ReceiptGenerator) -> None:
        self._tax_calculator = tax_calculator
        self._payment_gateway_client = payment_gateway_client
        self._receipt_generator = receipt_generator

    def process_checkout(self, cart: Cart, currency: str) -> Receipt:
        total_with_tax: Decimal = self._tax_calculator.calculate_total_with_tax(cart, currency)
        payment_id: str = self._payment_gateway_client.charge(total_with_tax, currency, cart.reference)
        return self._receipt_generator.generate(payment_id, cart, total_with_tax)
